File: ios/RunningOS/Data/Repositories/RunningRepositoryImpl.swift
import Foundation
import Combine

/// 跑步数据 Repository 实现
final class RunningRepositoryImpl: RunningRepository {

    // MARK: - Properties

    private let userDefaults: UserDefaults
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    // MARK: - Init

    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()

        // 设置日期格式
        let dateFormatter = ISO8601DateFormatter()
        dateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let dateString = try container.decode(String.self)

            if let date = dateFormatter.date(from: dateString) {
                return date
            }

            return Date()
        }
    }

    // MARK: - Public Methods

    func getUserStats() -> AnyPublisher<UserStats, Never> {
        // TODO: 从本地存储或网络获取真实数据
        // 这里返回模拟数据
        let stats = UserStats(
            totalDistance: 2450.0,
            totalRuns: 156,
            totalMarathons: 42,
            avgPace: 5.5,
            activeDays: 142,
            personalBests: PersonalBests(
                fiveK: "19:42",
                tenK: "41:15",
                halfMarathon: "1:32:04",
                marathon: "3:15:30"
            )
        )

        return Just(stats)
            .eraseToAnyPublisher()
    }

    func getAllActivities() -> AnyPublisher<[RunningActivity], Never> {
        // TODO: 从数据库获取真实数据
        return Just([])
            .eraseToAnyPublisher()
    }

    func getLatestActivity() -> AnyPublisher<RunningActivity?, Never> {
        // TODO: 从数据库获取最新活动
        return Just(nil)
            .eraseToAnyPublisher()
    }

    func getAllRoutes() -> AnyPublisher<[Route], Never> {
        // 返回东京环线示例数据
        let tokyoLoop = Route(
            id: "tokyo_loop",
            name: "Tokyo Loop",
            coordinates: [
                Coordinate(latitude: 35.6930, longitude: 139.7020),
                Coordinate(latitude: 35.7010, longitude: 139.7130),
                Coordinate(latitude: 35.7070, longitude: 139.7280),
                Coordinate(latitude: 35.6980, longitude: 139.7420),
                Coordinate(latitude: 35.6860, longitude: 139.7440),
                Coordinate(latitude: 35.6740, longitude: 139.7340),
                Coordinate(latitude: 35.6670, longitude: 139.7180),
                Coordinate(latitude: 35.6650, longitude: 139.7010),
                Coordinate(latitude: 35.6710, longitude: 139.6860),
                Coordinate(latitude: 35.6830, longitude: 139.6760),
                Coordinate(latitude: 35.6960, longitude: 139.6810),
                Coordinate(latitude: 35.7050, longitude: 139.6920),
                Coordinate(latitude: 35.7060, longitude: 139.7030),
                Coordinate(latitude: 35.7010, longitude: 139.7120),
                Coordinate(latitude: 35.6930, longitude: 139.7020)
            ],
            distance: 15.2,
            type: .loop
        )

        return Just([tokyoLoop])
            .eraseToAnyPublisher()
    }

    func getRouteById(_ routeId: String) -> Route? {
        // TODO: 实现从存储中查询
        return nil
    }

    func saveActivity(_ activity: RunningActivity) {
        // TODO: 保存到本地数据库
    }

    func syncData() -> AnyPublisher<Void, Error> {
        // TODO: 实现网络同步
        return Just(())
            .setFailureType(to: Error.self)
            .eraseToAnyPublisher()
    }
}

File: android/app/src/main/java/com/windchaser/runningos/data/repository/RunningRepositoryImpl.kt
package com.windchaser.runningos.data.repository

import com.windchaser.runningos.data.local.ActivityDao
import com.windchaser.runningos.data.local.ActivityEntity
import com.windchaser.runningos.data.local.RouteEntity
import com.windchaser.runningos.data.local.RouteDao
import com.windchaser.runningos.domain.model.Coordinate
import com.windchaser.runningos.domain.model.PersonalBests
import com.windchaser.runningos.domain.model.Route
import com.windchaser.runningos.domain.model.RouteType
import com.windchaser.runningos.domain.model.RunningActivity
import com.windchaser.runningos.domain.model.UserStats
import com.windchaser.runningos.domain.repository.RunningRepository
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject

/**
 * 跑步数据 Repository 实现
 */
class RunningRepositoryImpl @Inject constructor(
    private val activityDao: ActivityDao,
    private val routeDao: RouteDao,
    private val gson: Gson
) : RunningRepository {

    override fun getUserStats(): Flow<UserStats> {
        // TODO: 实现从数据库聚合统计数据
        // 这里使用模拟数据
        return kotlinx.coroutines.flow.flowOf(
            UserStats(
                totalDistance = 2450.0,
                totalRuns = 156,
                totalMarathons = 42,
                avgPace = 5.5,
                activeDays = 142,
                personalBests = PersonalBests(
                    fiveK = "19:42",
                    tenK = "41:15",
                    halfMarathon = "1:32:04",
                    marathon = "3:15:30"
                )
            )
        )
    }

    override fun getAllActivities(): Flow<List<RunningActivity>> {
        return activityDao.getAllActivities().map { entities ->
            entities.map { it.toDomainModel() }
        }
    }

    override fun getLatestActivity(): Flow<RunningActivity?> {
        return activityDao.getLatestActivity().map { entity ->
            entity?.toDomainModel()
        }
    }

    override fun getAllRoutes(): Flow<List<Route>> {
        return routeDao.getAllRoutes().map { entities ->
            entities.map { it.toDomainModel(gson) }
        }
    }

    override suspend fun getRouteById(routeId: String): Route? {
        return routeDao.getRouteById(routeId)?.toDomainModel(gson)
    }

    override suspend fun saveActivity(activity: RunningActivity) {
        activityDao.insertActivity(activity.toEntity())
    }

    override suspend fun syncData(): Result<Unit> {
        return try {
            // TODO: 实现网络同步逻辑
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

// 扩展函数：Entity ↔ Domain Model 转换
private fun ActivityEntity.toDomainModel(): RunningActivity {
    return RunningActivity(
        id = id,
        name = name,
        distance = distance,
        duration = duration,
        pace = pace,
        calories = calories,
        timestamp = timestamp,
        route = null
    )
}

private fun RunningActivity.toEntity(): ActivityEntity {
    return ActivityEntity(
        id = id,
        name = name,
        distance = distance,
        duration = duration,
        pace = pace,
        calories = calories,
        timestamp = timestamp,
        routeId = route?.id
    )
}

private fun RouteEntity.toDomainModel(gson: Gson): Route {
    val coordinateType = object : TypeToken<List<Coordinate>>() {}.type
    val coordinates: List<Coordinate> = gson.fromJson(coordinatesJson, coordinateType)

    return Route(
        id = id,
        name = name,
        coordinates = coordinates,
        distance = distance,
        type = when (type) {
            "LOOP" -> RouteType.LOOP
            "OUT_AND_BACK" -> RouteType.OUT_AND_BACK
            "POINT_TO_POINT" -> RouteType.POINT_TO_POINT
            else -> RouteType.LOOP
        }
    )
}
